### PR TITLE
[testing] Support input samples where `self` is broadcasted.

### LINKED ISSUE #50747: `test_inplace_grad` and `test_variant_consistency_eager` have to be disabled while testing broadcasting semantics in `OpInfo` based tests
As per title. see https://github.com/pytorch/pytorch/pull/50667

cc @mruberry @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #50747

Reference #50006

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -78,8 +78,14 @@ def _check_helper(self, device, dtype, op, variant, check):
         if not op.supports_dtype(dtype, torch.device(device).type):
             self.skipTest(f"Skipped! {op.name} does not support dtype {str(dtype)}")
 
+        def is_inplace(variant):
+            return variant.__name__.endswith('_')
```
Comment (kshitij12345) [inline test/test_ops.py:82]: Is there a better way to check this?

Or should we trickle variant type and variant from the function that calls this helper?

https://github.com/pytorch/pytorch/blob/c4c77e2001ead838932cab61c5813c25c4f3e72e/test/test_ops.py#L141
https://github.com/pytorch/pytorch/blob/c4c77e2001ead838932cab61c5813c25c4f3e72e/test/test_ops.py#L161

Comment (mruberry) [reply test/test_ops.py:82]: The correct way to check this is by testing if the variant `is` the inplace variant acquired from the OpInfo.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -76,22 +76,27 @@ class SampleInput(object):
     """Represents sample inputs to a function."""
 
     # output_process_fn_grad is a function that modifies the output of op compatible with input
-    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad']
+    __slots__ = ['input', 'args', 'kwargs', 'output_process_fn_grad', 'broadcasts_self']
 
-    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None):
+    def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=None, broadcasts_self=False):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:81]: This approach is reasonable but I still think we should filter sample_inputs() by whether the samples will be used for inplace computations or not. However, let me get a few other opinions on this. Maybe this is a better approach.

File: test/test_ops.py (hunk 1 above)
Comment (kshitij12345) [reply test/test_ops.py:82]: Unfortunately that won't work as the `op.get_inplace()` is wrapped by `_get_safe_inplace`.

https://github.com/pytorch/pytorch/blob/c4c77e2001ead838932cab61c5813c25c4f3e72e/test/test_ops.py#L68-L73

Comment (mruberry) [reply test/test_ops.py:82]: Can you access it through the dunder wrapped attribute?

```
foo.__wrapped__ is op.get_inplace()
```

Comment (kshitij12345) [reply test/test_ops.py:82]: Yup it works with that. Thanks!!

```python
def is_inplace(variant):
    if hasattr(variant, "__wrapped__"):
      return variant.__wrapped__ is op.get_inplace()

    return variant is op.get_inplace()
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:81]: Sorry I neglected to update this, @kshitij12345. General consensus is that it'd be preferable to filter sample_inputs() by what's inplace for now. We may want to revise that decision later, however.

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:81]: IIUC,
We will have to update the signature for every `sample_inputs_*` with 
```
sample_inputs_*(op_info, device, dtype, requires_grad, is_inplace_variant=False)
```
?

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:81]: We could do that (and make is_inplace_variant kwarg-only). That would be the most discoverable thing.

But we don't have to. The operations for testing the inplace variant can request sample inputs by passing the for_inplace_variant (we can work on the name) kwarg like this:

```
try:
  samples = op.sample_inputs(..., for_inplace_variant=True)
except TypeError as te:
  samples = op.sample_inputs(...) 
```

Then functions which support this option can implement it as:

```
def sample_inputs_foo(..., *, for_inplace_variant=False)
```

A nicer (but more disruptive) solution would be to make sample_inputs take **kwargs, and then the functions that want to use "for_inplace_variant" or similar options can query for it from the kwarg dict.

 What would you think of an approach like that?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:81]: I like the second approach (though more disruptive). Will try that. 
Thanks!

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:81]: For `test_variant_consistency_eager`, 

https://github.com/pytorch/pytorch/blob/0d81528a470f4dfa57461d19d24ed9baba904caf/test/test_ops.py#L204-L242

We will have to put call for `sample_inputs` inside the variants loop. Since call to sample_input has to actually materialize all the tensors, this will result in some performance regression for this test. Let me know if that sounds acceptable.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:81]: That's a great point, and that function could probably use a small refactoring. I agree with you but think we can mitigate the damage:

```
samples = op.sample_inputs(device, dtype, requires_grad=op.supports_autograd)

for sample in samples:
  # Acquires variants (method variant, inplace variant, aliases)
  method = op.get_method()
  inplace = op.get_inplace()

  # list of all inplace ops: inplace variant + alias inplace variants if exist
  inplace_ops = [inplace, ]

  aliases = []
  for a_op in op.aliases:
    aliases.append(a_op.op)
    aliases.append(a_op.method_variant)
    aliases.append(a_op.inplace_variant)
    inplace_ops.append(a_op.inplace_variant)
  
  aliases = tuple(aliases)
  inplace_ops = tuple(v for v in inplace_ops if v is not None)
  variants = (v for v in (method, inplace) + aliases if v is not None)
```

What's weird about this is that the test is rebuilding its understand of the operator's variants on each sample. This is unnecessary since variants are sample-independent. We should lift that section out of the for loop and acquire the variants once upfront.

Now let's look at the rest of the function:

```
# Computes function forward and backward values
sample.input.grad = None
expected_forward = op(sample.input, *sample.args, **sample.kwargs)
expected_grad = None

if (op.supports_autograd and isinstance(expected_forward, torch.Tensor)):
  expected_forward.sum().backward()
  expected_grad = sample.input.grad

  for variant in variants:
    sample.input.grad = None
    cloned = clone_input_helper(sample.input) if variant in inplace_ops else sample.input
    variant_forward = variant(cloned, *sample.args, **sample.kwargs)
    self.assertEqual(expected_forward, variant_forward)

    if expected_grad is not None and (variant not in inplace_ops or op.supports_inplace_autograd):
      variant_forward.sum().backward()
      self.assertEqual(expected_grad, sample.input.grad)
```

I think we can get rid of the `for variant in variants` loop here by using a itertools.product(sample inputs, variants) after the sample input and variant acquisition at the start of the test. Then we can create a helper function that executes this test body. The helper function takes a variant, a sample input, and whether the operation is inplace or not (so it knows whether to perform the copy. Then the test works like this:

- the variants are identified
- sample inputs are acquired
- if the op has an inplace variant, inplace sample inputs are acquired
- an itertools product of sample inputs x non-inplace variants invokes the helper function
- an itertools product of inplace sample inputs x inplace variants invokes the helper function

Does that make sense? Looking forward to hearing your thoughts.

You are correct that this will still redundantly create sample inputs that are common to both the out-of-place and in-place variant. Which is unfortunate from a performance standpoint, and your idea of marking SampleInputs as safe-for-inplace would have handled this problem better. However I think this is an OK penalty (for now, anyway).

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:81]: I'll try to do that. Will ping here if I stumble into another block.

Thanks!

File: test/test_ops.py [hunk 3]
```diff
@@ -203,56 +209,66 @@ class TestCommon(JitCommonTestCase):
...
+
+        inplace_variants = tuple(v for v in inplace_ops if v is not None)
+        variants = tuple(v for v in (method, inplace) + aliases if v is not None)
+        outplace_variants = tuple(set(variants) - set(inplace_ops))
+
+        samples = op.sample_inputs(device, dtype, requires_grad=op.supports_autograd,
+                                   for_inplace_variant=False)
+
+        def _test_consistency_helper(samples, op_variants):
+            for sample in samples:
```
Comment (kshitij12345) [inline test/test_ops.py:235]: Reason for iterating over samples and then over variants (instead of taking product of samples and variants and calling the helper) is so that we can utilise the `expected_forward` and `expected_grad` for multiple variants (from the same variant group).

outplace_variants will include method, alias function, alias method (if available)
inplace_variants will include inplace method, and alias inplace_method  (if available)

Comment (mruberry) [reply test/test_ops.py:235]: You're right, this is a good refinement of the design.

Comment (kshitij12345) [reply test/test_ops.py:235]: This explains the reason for previous test following this structure :)

Comment (mruberry) [reply test/test_ops.py:235]: Yep, although the acquisition of the variants can still be done once upfront I think, right?

Comment (kshitij12345) [reply test/test_ops.py:235]: Yes.

**Comment by kshitij12345** [comment]
@mruberry, have fixed the merge conflicts. I think this is ready for another round. PTAL :) 

**Note**:
One thing to note is that this PR is volatile, in the sense that if a new `samle_input_*` is added with current signature (without **kwargs) then this PR will lead to errors as it will pass an invalid argument (so it should be rebased prior to landing and no new `sample_input_*` should be added in between).

Or should we add a try except (similar to the one mentioned above) temporarily and then remove once this signature becomes the norm?

```python
try:
  samples = op.sample_inputs(..., **kwargs)
except TypeError as te:
  samples = op.sample_inputs(...) 
```

Thanks!

File: test/test_ops.py [hunk 4]
```diff
@@ -102,7 +102,13 @@ def _check_helper(self, device, dtype, op, variant, check):
         if not op.supports_dtype(dtype, torch.device(device).type):
             self.skipTest(f"Skipped! {op.name} does not support dtype {str(dtype)}")
 
-        samples = op.sample_inputs(device, dtype, requires_grad=True)
+        def is_inplace(variant):
+            if hasattr(variant, "__wrapped__"):
+                return variant.__wrapped__ is op.get_inplace()
+            return variant is op.get_inplace()
+
+        samples = op.sample_inputs(device, dtype, requires_grad=True,
```
Comment (mruberry) [inline test/test_ops.py:110]: I think your proposal to wrap this in a try/except so not every sample_input needs to be updated to use **kwargs when this lands is a good idea.

File: test/test_ops.py [hunk 5]
```diff
@@ -230,56 +236,70 @@ def test_variant_consistency_eager(self, device, dtype, op):
...
+                    if expected_grad is not None and \
+                            (variant not in inplace_ops or op.supports_inplace_autograd):
+                        variant_forward.sum().backward()
+                        self.assertEqual(expected_grad, tensor.grad)
+
+        _test_consistency_helper(samples, outplace_variants)
+
+        if len(inplace_ops) > 0:
+            inplace_samples = op.sample_inputs(device, dtype, requires_grad=_requires_grad,
+                                               for_inplace_variant=True)
```
Comment (mruberry) [inline test/test_ops.py:301]: Can also wrap this in try/except to minimize logical merge conflicts

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -1809,39 +1809,46 @@ def sample_inputs_floor_divide(op_info, device, dtype, requires_grad):
...
+        yield SampleInput(make_arg((S, S)), args=(torch.randn(S, device=device) > 0, 10))
+        yield SampleInput(make_arg(()), args=(torch.randn((), device=device) > 0, 10))
+        yield SampleInput(make_arg(()), args=(torch.randn((), device=device) > 0, make_arg(())))
+        yield SampleInput(make_arg((S, S)), args=(torch.randn((), device=device) > 0, 10))
+
+        if not for_inplace_variant:
+            yield SampleInput(make_arg((S,)),
+                              args=(torch.randn(S, S, device=device) > 0, make_arg(())))
+            yield SampleInput(make_arg((S,)),
+                              args=(torch.randn(S, S, device=device) > 0, 10))
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:1846]: Why this change instead of using the bernoulli op?

**Review by mruberry — COMMENTED** [review]
Hey @kshitij12345! Overall this looks good, as usual.

I made a few comments; I like your suggestion to reduce logical merge conflicts by using a try/except. 

I think the OpInfos for masked_fill and masked_scatter can also be updated with this change, right?

https://github.com/pytorch/pytorch/blob/f16c82786048641d8e3e81076d5eb0a3d2583978/torch/testing/_internal/common_methods_invocations.py#L3039

https://github.com/pytorch/pytorch/blob/f16c82786048641d8e3e81076d5eb0a3d2583978/torch/testing/_internal/common_methods_invocations.py#L3048

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:1836]: `bernoulli` produces scalar tensor however we want (S, S) tensor.
https://github.com/pytorch/pytorch/blob/3d492b069767dd814fdf72129074fc0f4511f400/torch/testing/_internal/common_methods_invocations.py#L3908-L3909

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -253,13 +253,20 @@ def get_operator_variant(self):
...
         """
-        return self.sample_inputs_func(self, device, dtype, requires_grad)
+
+        # TODO: Remove the try/except once all operators have sample_inputs_func with
+        #       **kwargs in their signature.
+        try:
+            samples = self.sample_inputs_func(self, device, dtype, requires_grad, **kwargs)
+        except TypeError:
+            samples = self.sample_inputs_func(self, device, dtype, requires_grad)
+        return samples
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:269]: Have added the try/except here. (we don't need to add try/except in test_ops)

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:269]: @mruberry can you please see if this is ok? Will rebase accordingly then. Thanks!

File: test/test_ops.py [hunk 8]
```diff
@@ -234,56 +240,70 @@ def test_variant_consistency_eager(self, device, dtype, op):
             inplace_ops.append(a_op.inplace_variant)
         aliases = tuple(aliases)
 
-        inplace_ops = tuple(v for v in inplace_ops if v is not None)
-        variants = (v for v in (method, inplace) + aliases if v is not None)
+        inplace_variants = tuple(v for v in inplace_ops if v is not None)
+        variants = tuple(v for v in (method, inplace) + aliases if v is not None)
```
Comment (mruberry) [inline test/test_ops.py:244]: Nice fix

File: torch/testing/_internal/common_methods_invocations.py (hunk 7 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:269]: Yep, this seems fine

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1853]:
```
('masked_fill', (M,), (torch.BoolTensor(M, M).bernoulli_(), 10), 'broadcast_lhs'),
```

I guess I mean where did this case go? We don't have to worry about it in this PR.

**Review by mruberry — APPROVED** [review]
Cool! And thanks for including the fix for the variant generator. If you rebase this now I think we can get it landed, @kshitij12345.

I made one comment, but we don't need to worry about in this PR even if it is an issue.

File: torch/testing/_internal/common_methods_invocations.py (hunk 6 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:1963]: Ah, replaced it with ( semantically both are doing the same thing),
```python
SampleInput(make_arg((S,)),
            args=(torch.randn(S, S, device=device) > 0, 10)
```

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1963]: OK

**Comment by kshitij12345** [comment]
With this, we might want to update/add entry to mention this new change at https://github.com/pytorch/pytorch/wiki/Writing-tests-in-PyTorch-1.8 and https://github.com/pytorch/pytorch/issues/54261 (OpInfo porting tracker issue)

**Comment by mruberry** [comment]
Hmm... I was hoping to land this, but the ROCm build failure is worrying. @kshitij12345, would you take a look?
